Set Socket priority using SO_PRIORITY

axkar opened the pull request:
Set socket priority using SO_PRIORITY.

This code update is essential for passing the Security Level 1 tests. The main challenge addressed by this update is the generation of significant traffic to create congestion (overload) for the switch. The previous issue was that traffic originating from the CPU (via the dsa0 port) consistently had priority zero (the lowest). Consequently, the switch could not respond to the PROFINET controller, leading to a loss of communication with the switch. The switch was expected to communicate with the controller by sending messages with higher priority.

The PR ensures that all DCP frames generated by the P-Net stack have elevated priority compared to normal traffic. Not higher than 6, to not cause interruptions on level 7, where MRP and other network control protocols reside.

- For external switch ports, using the 'dcb app' command to configure the ports to pass all PROFINET packets (ethertype 0x8892) on priority 6 was sufficient.
- For the internal port 'dsa0', which transmits messages from the CPU, the same principle cannot be directly applied. Therefore, the application needs to create a socket and set the socket priority to priority 6 using the 'setsockopt' and 'SO_PRIORITY' option.

troglobit reviewed (changes requested):
This change looks good, but you need to describe *why* the change is made.  The summary line is OK, but there's no need to repeat what we can see in the code diff.  

Something along the lines of: "Ensure all DCP frames generated by the P-Net stack have elevated priority compared to normal traffic.  Not higher than 6, to not cause interruptions on level 7, where MRP and other network control protocols reside."

Maybe also quote the PROFINET standard, citing the QoS requirements for conformance.  But that should possibly go in the code itself as a comment.

troglobit reviewed (changes requested):
Looks much better, a good comment in the code clearly describing the why this the sockopt is needed, followed by a well written commit message.  Well done!

Only one thing left: formatting!  Please consider the following seven golden rules https://cbea.ms/git-commit/#seven-rules, in particular line lengths.

Commit:
Set socket priority to level 6 using SO_PRIORITY

During Security Level 1 testing, priority issues were discovered when
executing tests under high load and traffic congestion caused by garbage
packets. Specifically, DCP messages from the PROFINET device/switch were
not reaching the PROFINET Controller, resulting in communication loss.

To address this, we elevated the priority of PROFINET packets for the
entire device, as dictated by the PROFINET standard, distinguishing
between external and internal ports:

  - External Ports: For packets with ethertype 0x8892, we used the
  'dcb app' command in the PROFINET main() function to set the priority
  to level 6. This configuration ensures that all external traffic
  ingressing the ports with ethertype 0x8892 egresses the corresponding
  output port with priority 6.

  - Internal Ports (CPU originated traffic): For packets originating
  inside the CPU and transmitting through the internal port/interface
  connected to the CPU, we set the socket priority using setsockopt
  and the SO_PRIORITY option. The 'dcb app' approach is not feasible
  for internal traffic due to:

    - The kernel not knowing the name of the port on the NC/bridge side
      connected to the CPU port/bus.

    - The mdio register way gives just a single bit (single option) to
      select. It would be hard or impossible to expand the functionality
      to other protocols.

The following changes were made to implement the internal port priority
handling:

  - Added a setsockopt call to set SO_PRIORITY to 6 for the socket in
  the pnal_eth_init function, ensuring DCP frames have elevated priority
  and preventing interruptions on level 7.

  - Adjusted socket handling by setting the priority immediately after
  socket creation and before other configurations.

axkar commented:
Hope it looks better now :)

troglobit reviewed (approved):
Perfect! :1st_place_medal:
